src/main/java/org/opensearch/knn/plugin/search/KNNConcurrentSearchRequestDecider.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.plugin.search;
+
+import lombok.EqualsAndHashCode;
+import org.opensearch.index.IndexSettings;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.query.KNNQueryBuilder;
+import org.opensearch.search.deciders.ConcurrentSearchDecision;
+import org.opensearch.search.deciders.ConcurrentSearchRequestDecider;
+
+import java.util.Optional;
+
+/**
+ * Decides if the knn query uses concurrent segment search
+ * As of 2.17, this is only used when
+ *  - "index.search.concurrent_segment_search.mode": "auto" or
+ *  - "search.concurrent_segment_search.mode": "auto"
+ *
+ *  Note: the class is not thread-safe and a new instance needs to be created for each request
```

Comment:
Why not just making this one singleton? Is any state being modified?

Reply:
Unfortunately we can't, The plugin gives a factory extension and creates a new object which maintains state for a request, So returning same object for all requests can cause race conditions with unpredictable results.

Reply:
@jmazanec15 the decider Object is at per request level. So, if it is singleton then how will be the decision made?

Reply:
This decider object does not have any state. So the factory could return the same object every time.

Reply:
but Im fine creating a new object - seems like it could be risky to share if state is added in future.

Reply:
No the object has the state. If you see getDeciderObject is returning the state which was decided on the evaluateQuery function.

Reply:
Resolved with @jmazanec15 offline 

Here is the link which helps understand why its per request. https://github.com/opensearch-project/OpenSearch/blob/4802d0d744cfd19827ab61c0807f557b768bd1aa/server/src/main/java/org/opensearch/search/DefaultSearchContext.java#L948